source/utils/CMakeLists.txt

```diff
@@ -1,10 +1,13 @@
 # Copyright (C) 2024 Intel Corporation
 # SPDX-License-Identifier: MIT
 
-add_library(utils
-  STATIC
-  "logging.h"
-  "logging.cpp"
-)
+set(logging_SOURCES logging.h logging.cpp)
+
+if(NOT spdlog_FOUND)
+       add_library(utils STATIC ${logging_SOURCES})
+       target_include_directories(utils PRIVATE ${CMAKE_SOURCE_DIR}/third_party/spdlog_headers/)
+else()
+       add_library(utils ${logging_SOURCES})
+       target_link_libraries(utils spdlog::spdlog)
+endif()
```

Comment:
```
target_sources(utils PRIVATE
  "logging.h"
  "logging.cpp"
)
if(NOT spdlog_FOUND)
       target_include_directories(utils PRIVATE ${CMAKE_SOURCE_DIR}/third_party/spdlog_headers/)
else()
       target_link_libraries(utils PRIVATE spdlog::spdlog)
endif()
```
A few notes here:
- it can be problematic if the library type of `utils` is changed depending on such `if` condition
- please keep the linkage type consistent `PRIVATE` vs `PUBLIC`. Not sure which one is supposed to be here, depends on if `spdlog.h` headers are propagated in the `logging.h`. Didn't check that one
- I know the previous implementation was using `add_library(STATIC)`, but it might be more flexible if you make it an `OBJECT` type. In both cases anyway `PIC` might need to be added.

Reply:
Okay, I'll start with compiled languages and god forbid build systems are not my area of expertise.

I do believe I've addressed PRIVATE x PUBLIC consistency. I've added POSITION_INDEPENDENT_CODE to the system spdlog path (but I'd assume I should look for the build flags passed and add it there only conditionally?).

Would you have patience and some more tips how to address the first point? And refactoring to the object type can go to another PR?

Anyhow, thanks a lot for the feedback!

Reply:
> I've added POSITION_INDEPENDENT_CODE to the system spdlog path (but I'd assume I should look for the build flags passed and add it there only conditionally?).

The approach is perfect, just move it outside the conditional. With my discussion with CMake folks, they said there is virtually no downside to adding it and it can alleviate some headaches down the line. 

> I do believe I've addressed PRIVATE x PUBLIC consistency

Because of the includes, these should be PUBLIC
https://github.com/oneapi-src/level-zero/blob/fd459f28041c176df23a0b7b791ff20a7689d237/source/utils/logging.h#L19-L20

To break it down, in order to call `#include "utils/logging.h"` you need to be able to `#include <spdlog.h>`, but that is only made available in `spdlog::spdlog`, so whatever target consumes `utils` must also have access to `spdlog::spdlog`, hence the `PUBLIC`.

Side-note, you should also fix the consumer part to make sure it is `PRIVATE` there
https://github.com/oneapi-src/level-zero/blob/fd459f28041c176df23a0b7b791ff20a7689d237/source/CMakeLists.txt#L48-L49

> Would you have patience and some more tips how to address the first point? And refactoring to the object type can go to another PR?

Yes, and of course. Other than the comments above there are only 2 remaining points
- Moving the `add_library` outside of the conditional (making sure it's always `STATIC` for now)
- `s/CMAKE_SOURCE_DIR/PROJECT_SOURCE_DIR/`. Modern practice, chips away at `FetchContent` support.

Reply:
Thanks a lot for the detailed explanation here 🙏 ! I've updated the PR, and also added the SYSTEM_SPDLOG guard flag (I personally don't care much for the default if I can change/set it downstream).